src/deluge/dsp/master_compressor/master_compressor.cpp

```diff
@@ -30,14 +30,21 @@ MasterCompressor::MasterCompressor() {
 	wet = 1.0;
 }
 
-void MasterCompressor::render(StereoSample* buffer, uint16_t numSamples) {
+void MasterCompressor::render(StereoSample* buffer, uint16_t numSamples, int32_t masterVolumeAdjustmentL,
+                              int32_t masterVolumeAdjustmentR) {
 
 	StereoSample* thisSample = buffer;
 	StereoSample* bufferEnd = buffer + numSamples;
 	if (compressor.getThresh() < -0.001) {
+		double adjustmentL = (masterVolumeAdjustmentL) / 4294967296.0; //  *2.0 is <<1 from multiply_32x32_rshift32
+		double adjustmentR = (masterVolumeAdjustmentR) / 4294967296.0;
+		if (adjustmentL < 0.000001)
+			adjustmentL = 0.000001;
+		if (adjustmentR < 0.000001)
+			adjustmentR = 0.000001;
 		do {
-			double l = thisSample->l / 2147483648.0;
-			double r = thisSample->r / 2147483648.0;
+			double l = thisSample->l / 2147483648.0 / adjustmentL;
+			double r = thisSample->r / 2147483648.0 / adjustmentR;
```

Comment:
Instead of converting the master volume adjustments to floats, I think we can apply it directly to `thisSample->l` and `thisSample->r` to get everything done in integers (potentially requires `multiply_32x32_rshift32`?)

Reply:
At this stage, floating point is required to process without clipping with masterVolumeAdjustment taken into account. Additionally, some of the operations have been modified to use integers.